### PR TITLE
Refactoring on all Framework

### PR DESCRIPTION
**Pull request by devGnode**

### CODE REVIEW

File: src/main/java/fr/zenity/desbugs/Enum/DesbugsPage.java [hunk 1]
```diff
@@ -12,28 +16,31 @@
...
+    ENTREPRISE("/enterprises/verify/%s"),
     ACCOUNT_GENERAL("/account/general"),
     ACCOUNT_IDENTIFY("/account/identifier"),
     ACCOUNT_SECURITY("/account/security");
 
     private String pageUrl;
 
-    public String getUrl(){
+    public String getUrl(String ... args){
+        AtomicReference<String> urlA = new AtomicReference<>(null);
```
Comment (l-bachmann) [inline src/main/java/fr/zenity/desbugs/Enum/DesbugsPage.java]: Why use AtomicReference ?

File: src/main/java/fr/zenity/desbugs/Enum/DesbugsPage.java [hunk 2]
```diff
@@ -49,13 +56,24 @@ public String getUrl(){
...
-    }
 
+        if( args.length > 0 ){
+
+            Arrays.asList(args)
+            .stream()
+            .forEach(elementValue->{
+                urlA.set( String.format( urlA.get(), elementValue ) );
+            });
+        }
```
Comment (l-bachmann) [inline src/main/java/fr/zenity/desbugs/Enum/DesbugsPage.java]: Currently, the urls have only one argument ; why use a stream ?

File: src/main/java/fr/zenity/desbugs/Enum/DesbugsPage.java [hunk 3]
```diff
@@ -49,13 +56,24 @@ public String getUrl(){
...
+                break;
         }
-    }
 
+        if( args.length > 0 ){
+
+            Arrays.asList(args)
+            .stream()
+            .forEach(elementValue->{
+                urlA.set( String.format( urlA.get(), elementValue ) );
```
Comment (l-bachmann) [inline src/main/java/fr/zenity/desbugs/Enum/DesbugsPage.java]: In this case, I think than you erase the url value with the arg value.
We want to replace the regex with the arg value instead of that.

File: src/main/java/fr/zenity/desbugs/Enum/UrlEnvironment.java [hunk 4]
```diff
@@ -38,23 +38,33 @@ public static UrlEnvironment initCustom(String env){
...
     }
 
-    public void setUrls(String urlsProperties){
-        Map<String, String> urls = Splitter.on(",").withKeyValueSeparator("=").split(urlsProperties.replaceAll("\\s+", ""));
-        urlLanding = urls.get("urlLanding");
-        urlApp = urls.get("urlApp");
+    public void setUrls(String urlsProperties) {
+
+        if(urlsProperties.equals("null"))
+            urlLanding = urlApp = null;
```
Comment (l-bachmann) [inline src/main/java/fr/zenity/desbugs/Enum/UrlEnvironment.java]: I think, the urls should never equal "null". Should we return an error in this case ?

File: src/main/java/fr/zenity/desbugs/Enum/UrlEnvironment.java [hunk 5]
```diff
@@ -38,23 +38,33 @@ public static UrlEnvironment initCustom(String env){
...
+        else if (!urlsProperties.equals("null") && Pattern.compile("(\\w+\\s*\\=\\s*([^\\,]+)?(\\,|))+").matcher(urlsProperties).find()){
```
Comment (l-bachmann) [inline src/main/java/fr/zenity/desbugs/Enum/UrlEnvironment.java]: The "null" value is already handle ; the "!urlsProperties.equals("null")" check is not necessary here.

File: src/main/java/fr/zenity/desbugs/Enum/UrlEnvironment.java [hunk 6]
```diff
@@ -38,23 +38,33 @@ public static UrlEnvironment initCustom(String env){
...
+
+            Map<String, String> urls = Splitter.on(",")
+                                        .withKeyValueSeparator("=")
+                                        .split(urlsProperties.replaceAll("\\s+", ""));
+            urlLanding  = urls.get("urlLanding");
+            urlApp      = urls.get("urlApp");
+        }
+        else
+            urlApp = urlsProperties;
```
Comment (l-bachmann) [inline src/main/java/fr/zenity/desbugs/Enum/UrlEnvironment.java]: I am not sure we should do that because we can set every wrong format in this parameter.
What is the purpose here ? Why is the urlLanding not set ?

File: src/main/java/fr/zenity/desbugs/PagesObjects/AccountGeneralPage.java [hunk 7]
```diff
@@ -8,7 +8,7 @@
 
     public AccountGeneralPage(WebDriver driver) {
         super(driver);
-        init(DesbugsPage.ACCOUNT_GENERAL, container);
+        init(DesbugsPage.RANKING, container);
```
Comment (l-bachmann) [inline src/main/java/fr/zenity/desbugs/PagesObjects/AccountGeneralPage.java]: RANKING is present in every PageObject class. I think it will be dificult to just undo your changes for these files. If you want, just delete the constructor  like you did before and I will manage the rebase.

File: src/test/java/fr/zenity/desbugs/runners/TestRunner.java [hunk 8]
```diff
@@ -3,9 +3,10 @@
 import io.cucumber.testng.CucumberOptions;
 
 @CucumberOptions(
-        features = "src/test/resources/features",
-        glue = {"fr.zenity.desbugs.steps","fr/zenity/desbugs/PagesObjects"},
-        tags = "",
+        features = "./src/test/resources/features",
+        strict = true,
+        glue = {"fr.zenity.desbugs.step_definitions","fr/zenity/desbugs/PagesObjects"},
```
Comment (l-bachmann) [inline src/test/java/fr/zenity/desbugs/runners/TestRunner.java]: Why did you rename the repository ?

File: src/main/java/fr/zenity/desbugs/Enum/DesbugsPage.java (hunk 2 above)
Comment (devGnode) [reply src/main/java/fr/zenity/desbugs/Enum/DesbugsPage.java]: Because stream is better than a simple loop, show stream for java8, use a loop or stream its same, now on java 8 its better to use stream than loop

File: src/main/java/fr/zenity/desbugs/Enum/DesbugsPage.java (hunk 3 above)
Comment (devGnode) [reply src/main/java/fr/zenity/desbugs/Enum/DesbugsPage.java]: No

File: src/main/java/fr/zenity/desbugs/Enum/UrlEnvironment.java (hunk 6 above)
Comment (devGnode) [reply src/main/java/fr/zenity/desbugs/Enum/UrlEnvironment.java]: read the algorithmique

File: src/test/java/fr/zenity/desbugs/runners/TestRunner.java (hunk 8 above)
Comment (devGnode) [reply src/test/java/fr/zenity/desbugs/runners/TestRunner.java]: is it just an tes, because the gherkin been not recognized, and this modification has no real impact on the framework

File: src/main/java/fr/zenity/desbugs/Enum/DesbugsPage.java (hunk 1 above)
Comment (devGnode) [reply src/main/java/fr/zenity/desbugs/Enum/DesbugsPage.java]: En anglais cela va être compliquer, AtomicRéference est du qu'en java il y a pas la portabilités des variables comme en javascript la seul mannière accèder à la modification d'une valeur dans une lambda soit utiliser un Array ou alors object AtomicRéferences

File: src/main/java/fr/zenity/desbugs/Enum/DesbugsPage.java (hunk 2 above)
Comment (l-bachmann) [reply src/main/java/fr/zenity/desbugs/Enum/DesbugsPage.java]: I maintain than I think it is not the best solution in this specific case but we can keep it like that at the moment.

File: src/main/java/fr/zenity/desbugs/Enum/DesbugsPage.java (hunk 3 above)
Comment (l-bachmann) [reply src/main/java/fr/zenity/desbugs/Enum/DesbugsPage.java]: My mistake

File: src/main/java/fr/zenity/desbugs/Enum/UrlEnvironment.java (hunk 6 above)
Comment (l-bachmann) [reply src/main/java/fr/zenity/desbugs/Enum/UrlEnvironment.java]: On this website, there will be always be 2 urls so the properties file should contains both.
If one is missing, we should not run the test and throw an error.
